File: Blazor.Canvas/Contexts/Context2D.cs
using Microsoft.JSInterop;
using System.Text.RegularExpressions;

namespace Blazor.Canvas.Contexts;

/// <summary>
/// Provides the 2D rendering context for the drawing surface of a <c>&lt;canvas&gt;</c> element.<br />
/// It is used for drawing shapes, text, images, and other objects.
/// </summary>
public class Context2D
{
	public BlazorCanvas Canvas { get; }

	protected string Id { get; }

	private IJSInProcessRuntime JS { get; }

	public Context2D(IJSRuntime js, string id, BlazorCanvas ownerCanvas)
	{
		Id = id;
		JS = (IJSInProcessRuntime)js;
		Canvas = ownerCanvas;
	}

	/* Properties */

	public async ValueTask<string> FillStyleAsync() => await JS.InvokeAsync<string>("getFillStyle", Id);
	public async Task FillStyleAsync(string style) => await JS.InvokeVoidAsync("setFillStyle", Id, style);

	public async ValueTask<double> GlobalAlphaAsync() => await JS.InvokeAsync<double>("getGlobalAlpha", Id);
	public async Task GlobalAlphaAsync(double alpha) => await JS.InvokeVoidAsync("setGlobalAlpha", Id, alpha);

	public async ValueTask<CompositeOperation> GlobalCompositeOperationAsync()
	{
		string jsString = await JS.InvokeAsync<string>("getGlobalCompositeOperation", Id);
		return Enum.Parse<CompositeOperation>(jsString.Replace("-", ""), true);
	}

	public async Task GlobalCompositeOperationAsync(CompositeOperation operation)
	{
		var regex = new Regex(@"
			(?<=[A-Z])(?=[A-Z][a-z]) |
			(?<=[^A-Z])(?=[A-Z]) |
			(?<=[A-Za-z])(?=[^A-Za-z])", RegexOptions.IgnorePatternWhitespace);

		string jsString = regex.Replace(Enum.GetName(operation)!, "-").ToLower();
		await JS.InvokeVoidAsync("setGlobalCompositeOperation", Id, jsString);
	}

	public async ValueTask<LineCap> LineCapAsync() => Enum.Parse<LineCap>(await JS.InvokeAsync<string>("getLineCap", Id), true);
	public async Task LineCapAsync(LineCap cap) => await JS.InvokeVoidAsync("setLineCap", Id, Enum.GetName(cap)!.ToLower());

	public async ValueTask<double> LineDashOffsetAsync() => await JS.InvokeAsync<double>("getLineDashOffset", Id);
	public async Task LineDashOffsetAsync(double offset) => await JS.InvokeVoidAsync("setLineDashOffset", Id, offset);

	public async ValueTask<LineJoin> LineJoinAsync() => Enum.Parse<LineJoin>(await JS.InvokeAsync<string>("getLineJoin", Id), true);
	public async Task LineJoinAsync(LineJoin join) => await JS.InvokeVoidAsync("setLineJoin", Id, Enum.GetName(join)!.ToLower());

	public async ValueTask<double> LineWidthAsync() => await JS.InvokeAsync<double>("getLineWidth", Id);
	public async Task LineWidthAsync(double width) => await JS.InvokeVoidAsync("setLineWidth", Id, width);

	public async ValueTask<string> StrokeStyleAsync() => await JS.InvokeAsync<string>("getStrokeStyle", Id);
	public async Task StrokeStyleAsync(string style) => await JS.InvokeVoidAsync("setStrokeStyle", Id, style);

	/* Methods */

	public void BeginPath()
		=> JS.InvokeVoid("doBeginPath", Id);

	public void LineTo(double x, double y)
		=> JS.InvokeVoid("doLineTo", Id, x, y);

	public void MoveTo(double x, double y)
		=> JS.InvokeVoid("doMoveTo", Id, x, y);

	public void Stroke()
		=> JS.InvokeVoid("doStroke", Id);

	/* Async methods */

	public async Task ArcAsync(double x, double y, double radius, double startAngle, double endAngle, bool counterclockwise = false)
		=> await JS.InvokeVoidAsync("doArc", Id, x, y, radius, startAngle, endAngle, counterclockwise);

	public async Task BeginPathAsync()
		=> await JS.InvokeVoidAsync("doBeginPath", Id);

	public async Task ClearRectAsync(double x, double y, double width, double height)
		=> await JS.InvokeVoidAsync("doClearRect", Id, x, y, width, height);

	public async Task ClosePathAsync()
		=> await JS.InvokeVoidAsync("doClosePath", Id);

	public async Task DrawImageAsync(string jsVariableName, double x, double y)
		=> await JS.InvokeVoidAsync("doDrawImage", Id, jsVariableName, x, y);

	public async Task DrawImageAsync(string jsVariableName, double x, double y, double width, double height)
		=> await JS.InvokeVoidAsync("doDrawImage2", Id, jsVariableName, x, y, width, height);

	public async Task EllipseAsync(double x, double y, double radiusX, double radiusY, double rotation, double startAngle, double endAngle, bool counterclockwise = false)
		=> await JS.InvokeVoidAsync("doEllipse", Id, x, y, radiusX, radiusY, rotation, startAngle, endAngle, counterclockwise);

	public async Task FillAsync()
		=> await JS.InvokeVoidAsync("doFill", Id);

	public async Task FillRectAsync(double x, double y, double width, double height)
		=> await JS.InvokeVoidAsync("doFillRect", Id, x, y, width, height);

	public async ValueTask<ImageData> GetImageDataAsync(double x, double y, double width, double height)
		=> await JS.InvokeAsync<ImageData>("getImageData", Id, x, y, width, height);

	public async ValueTask<double[]> GetLineDashAsync()
		=> await JS.InvokeAsync<double[]>("getLineDash", Id);

	public async Task LineToAsync(double x, double y)
		=> await JS.InvokeVoidAsync("doLineTo", Id, x, y);

	public async Task MoveToAsync(double x, double y)
		=> await JS.InvokeVoidAsync("doMoveTo", Id, x, y);

	public async Task PutImageDataAsync(ImageData data, double x, double y)
		=> await JS.InvokeVoidAsync("putImageData", Id, data.Data, data.Width, data.Height, data.ColorSpace, x, y);

	public async Task RectAsync(double x, double y, double width, double height)
		=> await JS.InvokeVoidAsync("doRect", Id, x, y, width, height);

	public async Task RestoreAsync()
		=> await JS.InvokeVoidAsync("doRestore", Id);

	public async Task RotateAsync(double angle)
		=> await JS.InvokeVoidAsync("doRotate", Id, angle);

	public async Task SaveAsync()
		=> await JS.InvokeVoidAsync("doSave", Id);

	public async Task SetLineDashAsync(params double[] segments)
		=> await JS.InvokeVoidAsync("setLineDash", Id, segments);

	public async Task StrokeAsync()
		=> await JS.InvokeVoidAsync("doStroke", Id);

	public async Task TranslateAsync(double x, double y)
		=> await JS.InvokeVoidAsync("doTranslate", Id, x, y);
}

File: Blazor.Canvas/wwwroot/context2D.js
/* Classes */

class DotNetSerializableImageData {
	constructor(imageData) {
		this.data = new Uint8Array(imageData.data);
		this.width = imageData.width;
		this.height = imageData.height
		this.colorSpace = imageData.colorSpace;
	}
}

/* Properties */

function getFillStyle(id) { return window[id].fillStyle; }
function getGlobalAlpha(id) { return window[id].globalAlpha; }
function getGlobalCompositeOperation(id) { return window[id].globalCompositeOperation; }
function getLineCap(id) { return window[id].lineCap; }
function getLineDashOffset(id) { return window[id].lineDashOffset; }
function getLineJoin(id) { return window[id].lineJoin; }
function getLineWidth(id) { return window[id].lineWidth; }
function getStrokeStyle(id) { return window[id].strokeStyle; }

function setFillStyle(id, style) { window[id].fillStyle = style; }
function setGlobalAlpha(id, alpha) { window[id].globalAlpha = alpha; }
function setGlobalCompositeOperation(id, operation) { window[id].globalCompositeOperation = operation; }
function setLineCap(id, cap) { window[id].lineCap = cap; }
function setLineDashOffset(id, offset) { window[id].lineDashOffset = offset; }
function setLineJoin(id, join) { window[id].lineJoin = join; }
function setLineWidth(id, width) { window[id].lineWidth = width; }
function setStrokeStyle(id, style) { window[id].strokeStyle = style; }

/* Methods */

function doArc(id, x, y, radius, startAngle, endAngle, counterclockwise) { window[id].arc(x, y, radius, startAngle, endAngle, counterclockwise); }
function doBeginPath(id) { window[id].beginPath(); }
function doClearRect(id, x, y, width, height) { window[id].clearRect(x, y, width, height); }
function doClosePath(id) { window[id].closePath(); }
function doEllipse(id, x, y, radiusX, radiusY, rotation, startAngle, endAngle, counterclockwise) { window[id].ellipse(x, y, radiusX, radiusY, rotation, startAngle, endAngle, counterclockwise); }
function doFill(id) { window[id].fill(); }
function doFillRect(id, x, y, width, height) { window[id].fillRect(x, y, width, height); }
function doLineTo(id, x, y) { window[id].lineTo(x, y); }
function doMoveTo(id, x, y) { window[id].moveTo(x, y); }
function doRect(id, x, y, width, height) { window[id].rect(x, y, width, height); }
function doRestore(id) { window[id].restore(); }
function doRotate(id, angle) { window[id].rotate(angle); }
function doSave(id) { window[id].save(); }
function doStroke(id) { window[id].stroke(); }
function doTranslate(id, x, y) { window[id].translate(x, y); }

function getLineDash(id) { return window[id].getLineDash(); }
function setLineDash(id, segments) { window[id].setLineDash(segments); }

function getImageData(id, x, y, width, height) {
	return new DotNetSerializableImageData(window[id].getImageData(x, y, width, height));
}

function putImageData(id, data, imageWidth, imageHeight, colorSpace, x, y) {
	window[id].putImageData(new ImageData(new Uint8ClampedArray(data), imageWidth, imageHeight, { colorSpace: colorSpace }), x, y);
}

function doDrawImage(id, jsVarName, x, y) { window[id].drawImage(window[jsVarName], x, y) }
function doDrawImage2(id, jsVarName, x, y, width, height) { window[id].drawImage(window[jsVarName], x, y, width, height); }
